docs/design: Initial schema blocks support documentation (#195)
hashicorp/terraform-plugin-framework: docs/design/blocks.md

resource "example_thing" "example" {
  some_attr = "abc123" # Example attribute/argument
}
```

While blocks are structural and defined without an equals sign (`=`), such as:

```terraform
resource "example_thing" "example" {
  some_block {
    # ... attributes/arguments ...
  }
}
```

Blocks can be repeated multiple times if the underlying schema type allows multiple declarations, such as:

```terraform
resource "example_thing" "example" {
  some_block {
    # ... attributes/arguments ...
  }

  some_block {
    # ... attributes/arguments ...
  }
}
```

To handle situations where the block configuration is not static, the configuration language added support for [`dynamic` block expressions](https://www.terraform.io/docs/language/expressions/dynamic-blocks.html). In contrast, working with attributes is relatively simpler as there are many expressions that can operate directly on any collection value, such as [`for` expressions](https://www.terraform.io/docs/language/expressions/for.html).

Blocks also have other limitations in Terraform, such as not supporting sensitivity. Instead, each nested attribute must implement sensitivity as appropriate.

### Terraform Plugin Protocol

> In terms of the protocol, all configuration information from the plugin is referred to as the schema. Prior to the recent version 6 of the protocol, only blocks were supported implementing complex schema types.

Reviewer: ```suggestion
In terms of the protocol, all configuration information from the plugin is referred to as the schema. Prior to the recent version 6 of the protocol, implementing complex schema types could only be done using blocks.
```

Perhaps also add a brief sentence pointing out the addition of `nested_type` to `Attribute` and what exactly this enables.

Author: Thanks! Pulled in that change and added a quick sentence there about `nested_type`.

	//
	// MaxItems can only be set for SchemaNestedBlockNestingModeList and
	// SchemaNestedBlockNestingModeSet. SchemaNestedBlockNestingModeSingle
	// can also set MinItems and MaxItems both to 1 to indicate that the
	// block is required to be set. All other SchemaNestedBlockNestingModes
	// must leave MaxItems set to 0.
	MaxItems int64
}
```

## Prior Implementations

### terraform-plugin-sdk

The previous framework provided the [`helper/schema.Resource` type](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema#Resource):

```go
// Resource represents a thing in Terraform that has a set of configurable
// attributes and a lifecycle (create, read, update, delete).
//
// The Resource schema is an abstraction that allows provider writers to
// worry only about CRUD operations while off-loading validation, diff
// generation, etc. to this higher level library.
//
// In spite of the name, this struct is not used only for terraform resources,
// but also for data sources. In the case of data sources, the Create,
// Update and Delete functions must not be provided.
type Resource struct {
    // ... other fields ...

>     // Schema is the schema for the configuration of this resource.

Reviewer 2: the whitespace seems off here

- Validating any constraints imposed by Terraform, such as lack of sensitivity.

## Proposals

### Blocks as New Nested Attribute Types

Blocks can be implemented within the existing `Attributes` handling inside `Attribute`.

For example:

```go
"example_block": {
	Attributes: ListNestedBlocks(map[string]Attribute{
		"example_attribute": {
			Type:     types.StringType,
			Required: true,
		},
	}, ListNestedBlocksOptions{}),
},
```

While this prevents the addition of a new field on `Attribute`, it can potentially conflate the two concepts and their supported functionality. Future enhancements that only apply to nested attributes may bleed into the overloaded schema definition. Any functionality within the framework would also need to differentiate the `Attributes` types by querying the nesting mode returned by the helper functions, rather than field existance.

### Blocks as New Attribute Field

Blocks can be implemented as a new field inside `Attribute`.

For example:

```go
"example_block": {
	Blocks: ListNestedBlocks(map[string]Attribute{
		"example_attribute": {
			Type:     types.StringType,
			Required: true,
		},
	}, ListNestedBlocksOptions{}),
},
```

> This allows the framework to differentiate between block and nested attributes by fields.

Reviewer 3: My question here is that I think this could allow providers to define incorrect schemas that we can't model with the protocol. For example, a ListNestedAttributes could contain a `tfsdk.Attribute` that has a `Blocks` property on it, which, if memory serves, we can't actually model in the protocol; attributes can only have attributes under them.

I think allowing this in the type system when it's not supported in the protocol may be confusing.
